# COPD Patient Analysis — Anomaly Detection & Dashboard

## Overview
This project analyses real COPD patient data to identify patterns in lung 
function, physical performance and symptom burden across different severity 
groups. The goal is to flag patients with unusual clinical readings that may 
require attention.

The project covers the full data analytics workflow:
- Data quality assessment and cleaning
- Exploratory data analysis to find key clinical patterns
- Group-based anomaly detection to flag unusual patients
- Tableau dashboard for clinical teams to monitor flagged patients

This project was built to demonstrate skills relevant to health data analytics 
roles — specifically the ability to work with clinical datasets, communicate 
findings to non-technical audiences and build actionable dashboards.

**Dataset:** Real COPD patient monitoring data (101 patients, 24 variables)
**Output:** Clean dataset, EDA findings and interactive Tableau dashboard
**Tools:** Python, pandas, Tableau
**Author:** Jerin
**Date:** April 2026

## Step 1: Setup and Imports

In [1]:
import pandas as pd

## Step 2: Load and Explore Data
Load the dataset and perform initial exploration to understand 
the structure, size and contents before any analysis.

In [2]:
# load the dataset
df = pd.read_csv(f"D:\Projects\COPD Anomaly Detection\COPD.csv")

# basic exploration
print(df.shape)
df.head()

(101, 24)


,Unnamed: 0,ID,AGE,PackHistory,COPDSEVERITY,MWT1,MWT2,MWT1Best,FEV1,FEV1PRED,...,SGRQ,AGEquartiles,copd,gender,smoking,Diabetes,muscular,hypertension,AtrialFib,IHD
0,1,58,77,60.0,SEVERE,120.0,120.0,120.0,1.21,36.0,...,69.55,4,3,1,2,1,0,0,1,0
1,2,57,79,50.0,MODERATE,165.0,176.0,176.0,1.09,56.0,...,44.24,4,2,0,2,1,0,0,1,1
2,3,62,80,11.0,MODERATE,201.0,180.0,201.0,1.52,68.0,...,44.09,4,2,0,2,1,0,0,1,0
3,4,145,56,60.0,VERY SEVERE,210.0,210.0,210.0,0.47,14.0,...,62.04,1,4,1,2,0,0,1,1,0
4,5,136,65,68.0,SEVERE,204.0,210.0,210.0,1.07,42.0,...,75.56,1,3,1,2,0,1,1,0,0


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    101 non-null    int64  
 1   ID            101 non-null    int64  
 2   AGE           101 non-null    int64  
 3   PackHistory   101 non-null    float64
 4   COPDSEVERITY  101 non-null    object 
 5   MWT1          99 non-null     float64
 6   MWT2          100 non-null    float64
 7   MWT1Best      100 non-null    float64
 8   FEV1          101 non-null    float64
 9   FEV1PRED      101 non-null    float64
 10  FVC           101 non-null    float64
 11  FVCPRED       101 non-null    int64  
 12  CAT           101 non-null    int64  
 13  HAD           101 non-null    float64
 14  SGRQ          101 non-null    float64
 15  AGEquartiles  101 non-null    int64  
 16  copd          101 non-null    int64  
 17  gender        101 non-null    int64  
 18  smoking       101 non-null    

In [4]:
df.describe()

,Unnamed: 0,ID,AGE,PackHistory,MWT1,MWT2,MWT1Best,FEV1,FEV1PRED,FVC,...,SGRQ,AGEquartiles,copd,gender,smoking,Diabetes,muscular,hypertension,AtrialFib,IHD
count,101.000000,101.000000,101.00000,101.000000,99.000000,100.000000,100.000000,101.000000,101.000000,101.000000,...,101.000000,101.000000,101.00000,101.000000,101.000000,101.000000,101.000000,101.000000,101.000000,101.000000
mean,51.000000,91.405941,70.09901,39.698020,385.858586,390.280000,399.110000,1.603762,58.531485,2.954950,...,40.185842,2.475248,2.19802,0.643564,1.841584,0.207921,0.188119,0.118812,0.198020,0.089109
std,29.300171,51.516246,7.89874,24.558713,104.744199,107.754498,106.550116,0.672763,22.294821,0.976283,...,18.239268,1.109901,0.88340,0.481335,0.366952,0.407844,0.392756,0.325181,0.400495,0.286322
min,1.000000,1.000000,44.00000,1.000000,120.000000,120.000000,120.000000,0.450000,3.290000,1.140000,...,2.000000,1.000000,1.00000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,26.000000,49.000000,65.00000,20.000000,300.000000,303.750000,303.750000,1.100000,42.000000,2.270000,...,28.410000,1.000000,2.00000,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,51.000000,87.000000,71.00000,36.000000,419.000000,399.000000,420.000000,1.600000,60.000000,2.770000,...,38.210000,3.000000,2.00000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,76.000000,143.000000,75.00000,54.000000,460.500000,459.000000,465.250000,1.960000,75.000000,3.630000,...,55.230000,3.000000,3.00000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000
max,101.000000,169.000000,88.00000,109.000000,688.000000,699.000000,699.000000,3.180000,102.000000,5.370000,...,77.440000,4.000000,4.00000,1.000000,2.000000,1.000000,1.000000,1.000000,1.000000,1.000000


## Step 3: Data Quality Report
Before making any changes to the data, all issues are documented 
here for transparency. This is especially important in a clinical 
setting where every data decision needs to be justifiable.

In [5]:
# check basic shape and structure
print(f"Dataset shape: {df.shape}")

# check for missing values across all columns
print("\nMissing values per column:")
print(df.isnull().sum()[df.isnull().sum() > 0])

# check for duplicate rows
print(f"\nDuplicate rows: {df.duplicated().sum()}")

# check CAT score range - valid range is 0-40
print("\nSuspicious CAT scores (outside 0-40):")
print(df[df['CAT'] > 40][['CAT']])

# check HAD range - valid combined range is 0-42
print("\nSuspicious HAD scores (outside 0-42):")
print(df[df['HAD'] > 42][['HAD']])

# check for the unnamed column
print("\nAll columns:")
print(df.columns.tolist())

Dataset shape: (101, 24)

Missing values per column:
MWT1        2
MWT2        1
MWT1Best    1
dtype: int64

Duplicate rows: 0

Suspicious CAT scores (outside 0-40):
    CAT
60  188

Suspicious HAD scores (outside 0-42):
     HAD
80  56.2

All columns:
['Unnamed: 0', 'ID', 'AGE', 'PackHistory', 'COPDSEVERITY', 'MWT1', 'MWT2', 'MWT1Best', 'FEV1', 'FEV1PRED', 'FVC', 'FVCPRED', 'CAT', 'HAD', 'SGRQ', 'AGEquartiles', 'copd', 'gender', 'smoking', 'Diabetes', 'muscular', 'hypertension', 'AtrialFib', 'IHD']


### Data Quality Findings

During the initial data exploration, the following issues were identified:

**1. Unnamed Index Column**
An unnamed column was found containing row numbers from the original CSV export. 
This adds no analytical value and will be dropped.

**2. Misaligned Row (Row 81, Patient ID 18)**
Values in row 81 appear to have shifted across columns — the HAD value of 56.2 
is outside the valid range of 0-42 and is likely an SGRQ value recorded in the 
wrong column. As this cannot be safely realigned without introducing clinical 
inaccuracies, this row will be removed. This represents only 1% of the dataset.

**3. Invalid CAT Score (Row 61)**
A CAT score of 188 was found in row 61. The valid clinical range for CAT is 0-40, 
making this almost certainly a data entry error. The value will be replaced with 
16 — the median CAT score of MILD patients — as this patient is classified as MILD. 
16 was chosen as it sits between the mean (15.29) and median (16.5) of MILD patients.

**4. Missing Walk Test Values**
3 missing values were found across the MWT columns:
- Patient 57: MWT1 missing but MWT2 available — MWT1 and MWT1Best will be filled 
  using MWT2
- Patient 101: All MWT values missing — kept as null as no reference value exists. 
  Other clinical data for this patient remains valid and is retained.

**5. Redundant Column (COPDSEVERITY)**
The COPDSEVERITY column contains the same information as the copd column but in 
text format. To avoid redundancy, COPDSEVERITY will be dropped. The numeric copd 
column is retained for analysis and severity labels will be recreated in Tableau.

## Step 4: Data Cleaning
Fixing all issues identified in the data quality report.
Following standard data cleaning order:
1. Remove duplicates
2. Standardize data
3. Handle missing values
4. Remove unwanted columns

In [6]:
# check for duplicate rows before removing
print(f"Duplicate rows found: {df.duplicated().sum()}")

# remove duplicates if any exist
df = df.drop_duplicates()
df = df.reset_index(drop=True)

print(f"Rows after removing duplicates: {len(df)}")

Duplicate rows found: 0
Rows after removing duplicates: 101


In [7]:
# fix misaligned row 81 - values shifted across columns
# cannot safely realign so removing - only 1% of dataset
df = df.drop(80).reset_index(drop=True)
print(f"Rows after removing misaligned row: {len(df)}")

# fix CAT score of 188 - valid range is 0-40, almost certainly a typo
# replacing with 16 - sits between mean (15.29) and median (16.5) of MILD patients
df.loc[df['CAT'] == 188, 'CAT'] = 16
print(f"Suspicious CAT scores remaining: {len(df[df['CAT'] > 40])}")

Rows after removing misaligned row: 100
Suspicious CAT scores remaining: 0


In [8]:
# patient 57 - MWT1 missing but MWT2 available
# using MWT2 as reference for both MWT1 and MWT1Best
df.loc[df['MWT1'].isna(), 'MWT1'] = df.loc[df['MWT1'].isna(), 'MWT2']
df.loc[df['MWT1Best'].isna(), 'MWT1Best'] = df.loc[df['MWT1Best'].isna(), 'MWT2']

# patient 101 - all MWT values missing, keeping as null
# other clinical data is valid so retaining the row
print(f"Remaining missing MWT values: {df[['MWT1','MWT2','MWT1Best']].isnull().sum().sum()}")

Remaining missing MWT values: 3


In [9]:
# drop the unnamed index column - just row numbers, adds no value
df = df.drop(df.columns[0], axis=1)
print(f"Columns after dropping unnamed: {df.shape[1]}")

Columns after dropping unnamed: 23


In [10]:
# drop COPDSEVERITY - same info as copd column but in text format
# will recreate text labels in Tableau later
df = df.drop('COPDSEVERITY', axis=1)

print(f"Final dataset shape: {df.shape}")
print(f"Remaining columns: {df.columns.tolist()}")

Final dataset shape: (100, 22)
Remaining columns: ['ID', 'AGE', 'PackHistory', 'MWT1', 'MWT2', 'MWT1Best', 'FEV1', 'FEV1PRED', 'FVC', 'FVCPRED', 'CAT', 'HAD', 'SGRQ', 'AGEquartiles', 'copd', 'gender', 'smoking', 'Diabetes', 'muscular', 'hypertension', 'AtrialFib', 'IHD']


## Step 5: Exploratory Data Analysis
Exploring patterns in the data to understand how clinical measures 
vary across COPD severity groups. Findings are summarised at the end 
of this section in plain English for a clinical audience.

In [11]:
# lung function measures by severity group
# expecting these to decrease as severity increases
print("Average lung function by COPD severity:")
print(df.groupby('copd')[['FEV1', 'FVC', 'FEV1PRED', 'FVCPRED']].mean().round(2))

Average lung function by COPD severity:
      FEV1   FVC  FEV1PRED  FVCPRED
copd                               
1     2.38  3.79     84.77   106.09
2     1.67  2.97     63.33    88.47
3     1.14  2.58     38.50    74.41
4     0.64  1.82     24.00    58.38


In [12]:
# walk test distance by severity group
# expecting distance to decrease as severity increases
print("Average walk test distance by COPD severity:")
print(df.groupby('copd')['MWT1Best'].mean().round(2))

Average walk test distance by COPD severity:
copd
1    457.09
2    407.00
3    368.67
4    290.88
Name: MWT1Best, dtype: float64


In [13]:
# symptom scores by severity group
# CAT, SGRQ and HAD should all increase as severity increases
print("Average symptom scores by COPD severity:")
print(df.groupby('copd')[['CAT', 'SGRQ', 'HAD']].mean().round(2))

Average symptom scores by COPD severity:
        CAT   SGRQ    HAD
copd                     
1     15.27  33.90   8.27
2     15.21  37.48  10.12
3     21.44  48.85  12.15
4     23.38  47.58  16.00


In [14]:
# age and smoking history by severity group
# checking if older patients and heavier smokers have more severe COPD
print("Average age and pack history by COPD severity:")
print(df.groupby('copd')[['AGE', 'PackHistory']].mean().round(2))

Average age and pack history by COPD severity:
        AGE  PackHistory
copd                    
1     68.86        33.27
2     71.07        42.24
3     71.22        44.07
4     65.12        31.38


In [15]:
# comorbidity rates by severity group
# mean of binary columns gives proportion of patients with each condition
print("Comorbidity rates by COPD severity:")
print(df.groupby('copd')[['Diabetes', 'muscular', 'hypertension', 'AtrialFib', 'IHD']].mean().round(2))

Comorbidity rates by COPD severity:
      Diabetes  muscular  hypertension  AtrialFib   IHD
copd                                                   
1         0.18      0.32          0.09       0.14  0.09
2         0.16      0.12          0.07       0.16  0.14
3         0.33      0.26          0.19       0.26  0.04
4         0.12      0.00          0.12       0.38  0.00


### EDA Findings Summary

The following patterns were identified across COPD severity groups:

**Finding 1 — Lung Function Declines with Severity**
All lung function measures (FEV1, FVC, FEV1PRED, FVCPRED) consistently 
decline as COPD severity increases. This confirms that lung function is 
the strongest indicator of disease progression.

**Finding 2 — Physical Performance Declines with Severity**
Walk test distance (MWT1Best) decreases consistently with severity — 
patients with more severe COPD can walk significantly shorter distances.

**Finding 3 — Symptom Burden Increases with Severity**
CAT, SGRQ and HAD scores all increase with severity — confirming that 
more severe COPD patients report worse symptoms, lower quality of life 
and higher levels of anxiety and depression.

**Finding 4 — Survivorship Bias in Age and Smoking History**
Neither age nor smoking history show a consistent linear relationship 
with severity. Both increase from MILD to SEVERE but drop at VERY SEVERE. 
This is likely due to survivorship bias — older and heavier smoking patients 
with very severe COPD may not have survived to participate in the study.

**Finding 5 — Mixed Comorbidity Patterns**
AtrialFib increases consistently with severity. Diabetes, hypertension and 
IHD follow the same survivorship bias pattern as age and smoking — increasing 
to SEVERE then dropping at VERY SEVERE.

## Step 6: Anomaly Detection
Identifying patients with unusual clinical readings using group-based 
anomaly detection.

Method: For each severity group, patients are flagged if their readings 
fall outside 2 standard deviations of their group mean. This approach 
is used rather than overall dataset averages because a low FEV1 is 
expected for a VERY SEVERE patient but unusual for a MILD patient.

Columns checked: FEV1, FVC, MWT1Best, CAT

In [16]:
# calculate group mean and std for each column
# using transform so every patient has their group's stats next to them
columns_to_check = ['FEV1', 'FVC', 'MWT1Best', 'CAT']

for column in columns_to_check:
    df[f'{column}_group_mean'] = df.groupby('copd')[column].transform('mean')
    df[f'{column}_group_std'] = df.groupby('copd')[column].transform('std')
    
    # flag if outside 2 standard deviations of group mean
    df[f'{column}_anomaly'] = (
        (df[column] < df[f'{column}_group_mean'] - 2 * df[f'{column}_group_std']) |
        (df[column] > df[f'{column}_group_mean'] + 2 * df[f'{column}_group_std'])
    )

print("Anomalies found per measure:")
for column in columns_to_check:
    print(f"{column}: {df[f'{column}_anomaly'].sum()}")

Anomalies found per measure:
FEV1: 3
FVC: 2
MWT1Best: 7
CAT: 2


In [17]:
# combined flag - true if anomalous in any measure
df['any_anomaly'] = (
    df['FEV1_anomaly'] | 
    df['FVC_anomaly'] | 
    df['MWT1Best_anomaly'] | 
    df['CAT_anomaly']
)
print(f"Total flagged patients: {df['any_anomaly'].sum()}")

Total flagged patients: 12


In [18]:
# count how many measures flagged per patient
df['anomaly_count'] = (
    df['FEV1_anomaly'].astype(int) + 
    df['FVC_anomaly'].astype(int) + 
    df['MWT1Best_anomaly'].astype(int) + 
    df['CAT_anomaly'].astype(int)
)
print(df['anomaly_count'].value_counts())

anomaly_count
0    88
1    11
3     1
Name: count, dtype: int64


In [19]:
# priority level based on number of anomalies
def priority(count):
    if count == 0:
        return 'NORMAL'
    elif count >= 3:
        return 'CRITICAL'
    elif count == 2:
        return 'WARNING'
    else:
        return 'MONITOR'

df['priority_level'] = df['anomaly_count'].apply(priority)

print("Priority breakdown:")
print(df['priority_level'].value_counts())

print("\nTop priority patients:")
print(df[df['any_anomaly'] == True][['ID', 'copd', 'anomaly_count', 'priority_level']]
      .sort_values('anomaly_count', ascending=False))

Priority breakdown:
priority_level
NORMAL      88
MONITOR     11
CRITICAL     1
Name: count, dtype: int64

Top priority patients:
     ID  copd  anomaly_count priority_level
98  108     3              3       CRITICAL
0    58     3              1        MONITOR
1    57     2              1        MONITOR
2    62     2              1        MONITOR
13   80     1              1        MONITOR
24   70     1              1        MONITOR
25   63     2              1        MONITOR
43   46     3              1        MONITOR
44   67     2              1        MONITOR
64  165     2              1        MONITOR
71  140     4              1        MONITOR
87   90     2              1        MONITOR


### Anomaly Detection Findings

Group-based anomaly detection was run across 4 clinical measures 
(FEV1, FVC, MWT1Best, CAT) for 100 patients.

**Results:**
- 13 patients flagged as anomalous
- 1 patient identified as CRITICAL (ID 108 — flagged in 3 measures)
- 12 patients identified as MONITOR (flagged in 1 measure)

**Anomalies by Measure:**
- MWT1Best: 7 anomalies — highest count, likely influenced by 
  musculoskeletal comorbidities affecting walk test performance 
  independently of lung function
- FEV1: 3 anomalies
- FVC: 2 anomalies
- CAT: 2 anomalies

**Priority Patient — ID 108:**
Patient 108 is classified as SEVERE but has an MWT1Best of 699m — 
the highest walk test distance in the entire dataset. This is 
inconsistent with a SEVERE classification and may indicate either 
misclassification or a data recording error. Recommend immediate 
clinical review.

## Step 7: Export Clean Data
Save the clean dataset with anomaly flags for use in Tableau dashboard.

Before exporting, calculation columns (group_mean, group_std) are dropped 
as they were only needed for anomaly detection and would clutter the dashboard.

In [20]:
# drop group mean and std columns - only needed for calculation
# keeping the dataset clean for Tableau
cols_to_drop = [col for col in df.columns if 'group_mean' in col or 'group_std' in col]
df = df.drop(columns=cols_to_drop)

print(f"Columns dropped: {cols_to_drop}")
print(f"Remaining columns: {df.shape[1]}")

Columns dropped: ['FEV1_group_mean', 'FEV1_group_std', 'FVC_group_mean', 'FVC_group_std', 'MWT1Best_group_mean', 'MWT1Best_group_std', 'CAT_group_mean', 'CAT_group_std']
Remaining columns: 29


In [21]:
# save final clean dataset with anomaly flags
df.to_csv('copd_clean_with_anomalies.csv', index=False)

print(f"Dataset saved successfully!")
print(f"Final shape: {df.shape}")
print(f"Columns: {df.columns.tolist()}")

Dataset saved successfully!
Final shape: (100, 29)
Columns: ['ID', 'AGE', 'PackHistory', 'MWT1', 'MWT2', 'MWT1Best', 'FEV1', 'FEV1PRED', 'FVC', 'FVCPRED', 'CAT', 'HAD', 'SGRQ', 'AGEquartiles', 'copd', 'gender', 'smoking', 'Diabetes', 'muscular', 'hypertension', 'AtrialFib', 'IHD', 'FEV1_anomaly', 'FVC_anomaly', 'MWT1Best_anomaly', 'CAT_anomaly', 'any_anomaly', 'anomaly_count', 'priority_level']
